### PR TITLE
Refactor `CombinedLoader` using pytrees

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Simplify the CombinedLoader implementation by using slightly modified pytrees from PyTorch.

A follow-up will look at their integration across the codebase.

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/pytorch/loops/fetchers.py [hunk 1]
```diff
@@ -45,14 +45,6 @@ def dataloader(self) -> Iterable:
             )
         return self._dataloader
 
-    @property
-    def loader_iters(self) -> Any:
```
Comment (carmocca) [inline src/lightning/pytorch/loops/fetchers.py:49]: This was unused after #16533

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 2]
```diff
@@ -239,28 +239,17 @@ def _prepare_dataloader(
         """This function handles the following functionalities:
 
         - Injecting a `DistributedDataSamplerWrapper` into the `DataLoader` if on a distributed environment
-        - Wrapping the datasets and samplers into fault-tolerant components
         - Wrapping the dataloader based on strategy-specific logic
         """
         if isinstance(dataloader, CombinedLoader):
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:244]: A follow-up will address how we interact with the `CombinedLoader` across the codebase

File: src/lightning/pytorch/trainer/supporters.py [hunk 3]
```diff
@@ -217,171 +95,91 @@ class CombinedLoader:
...
         if mode not in _supported_modes:
             raise ValueError(f"Unsupported mode {mode!r}, please select one of: {list(_supported_modes)}.")
+        self._loaders = loaders
+        self._loaders_flattened, self._loaders_spec = _tree_flatten(loaders)
 
-        self.loaders = loaders
-
-        datasets = apply_to_collection(
-            self.loaders, Iterable, getattr, "dataset", None, wrong_dtype=(Sequence, Mapping)
+        # FIXME(carlos): why do we even need this?
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/supporters.py:104]: This is something I want to explore, but not for this PR.

File: src/lightning/pytorch/trainer/supporters.py [hunk 4]
```diff
@@ -217,171 +95,91 @@ class CombinedLoader:
...
-
-    def __iter__(self) -> Any:
-        """Create and return an iterator, `CombinedLoaderIterator`, for the combined loader."""
-
-        # prevent `NotImplementedError` from PyTorch:
-        # https://github.com/pytorch/pytorch/blob/v1.9.0/torch/utils/data/dataloader.py#L541
-        def __getstate__patch__(*_: Any) -> Dict:
-            return {}
-
-        _BaseDataLoaderIter.__getstate__ = __getstate__patch__  # type: ignore[assignment]
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/supporters.py:285]: Unsure about whether this is still needed

File: tests/tests_pytorch/trainer/test_supporters.py [hunk 5]
```diff
@@ -265,26 +245,6 @@ def test_combined_loader_sequence_max_size_cycle():
...
-        ([*range(10), list(range(1, 20))], max, 19),
-        ([*range(10), {str(i): i for i in range(1, 20)}], min, 0),
-        ([*range(10), {str(i): i for i in range(1, 20)}], max, 19),
-        ({**{str(i): i for i in range(10)}, "nested": {str(i): i for i in range(1, 20)}}, min, 0),
-        ({**{str(i): i for i in range(10)}, "nested": {str(i): i for i in range(1, 20)}}, max, 19),
-        ({**{str(i): i for i in range(10)}, "nested": list(range(20))}, min, 0),
-        ({**{str(i): i for i in range(10)}, "nested": list(range(20))}, max, 19),
-    ],
-)
-def test_nested_calc_num_data(input_data, compute_func, expected_length):
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_supporters.py:281]: I tried to leave tests mostly untouched. But this one I didn't know what would be the equivalent version

File: src/lightning/pytorch/trainer/supporters.py [hunk 6]
```diff
@@ -217,171 +95,91 @@ class CombinedLoader:
...
+                    raise StopIteration
+                if self._mode == "max_size_cycle":
+                    # reset the consumed dataloader
+                    self._loader_iters[i] = iter(self._loaders_flattened[i])
+                    out[i] = next(self._loader_iters[i])
+                    continue
+                elif self._mode == "min_size":
+                    raise
+        if any(isinstance(v, _NoneSentinel) for v in out):
+            raise RuntimeError("Something went wrong! Please open an issue on GitHub.")
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/supporters.py:150]: I added this for sanity checking, but i'll probably remove it when it's all green

File: src/lightning/pytorch/trainer/supporters.py [hunk 7]
```diff
@@ -217,171 +95,91 @@ class CombinedLoader:
...
+    def __next__(self) -> Any:
+        n = len(self._loader_iters)
+        out = [_NoneSentinel()] * n  # values per iterator
+        for i in range(n):
+            try:
+                out[i] = next(self._loader_iters[i])
+            except StopIteration:
+                self._consumed[i] = True
+                if all(self._consumed):
+                    raise StopIteration
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/supporters.py:141]:
```suggestion
                    raise
```

File: src/lightning/pytorch/trainer/supporters.py (hunk 3 above)
Comment (Borda) [reply src/lightning/pytorch/trainer/supporters.py:104]: Is this then fixme (needs to be done before merge) or rather todo (can be done later?)

File: src/lightning/pytorch/utilities/_pytree.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+from typing import Any, List, Tuple
+
+from torch.utils._pytree import _get_node_type, LeafSpec, PyTree, SUPPORTED_NODES, tree_unflatten, TreeSpec
+
+
+def _is_leaf(pytree: PyTree) -> bool:
```
Comment (Borda) [inline src/lightning/pytorch/utilities/_pytree.py:6]: any dosctring?

**Review by Borda — APPROVED** [review]
Very nice simplification!

File: src/lightning/pytorch/utilities/_pytree.py (hunk 8 above)
Comment (carmocca) [reply src/lightning/pytorch/utilities/_pytree.py:6]: This is a copy of `torch.utils._pytree._is_leaf` and is not meant to be public. I added comments about the relevant modifications

File: src/lightning/pytorch/trainer/supporters.py (hunk 3 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/supporters.py]: Changed to TODO

File: src/lightning/pytorch/utilities/_pytree.py (hunk 8 above)
Comment (Borda) [reply src/lightning/pytorch/utilities/_pytree.py:6]: still docstring for future us, or if you want for me :P

Comment (carmocca) [reply src/lightning/pytorch/utilities/_pytree.py]: Done

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/CHANGELOG.md [hunk 9]
```diff
@@ -223,6 +224,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Removed the unused `lightning.pytorch.utilities.metrics.metrics_to_scalars` function ([#16681](https://github.com/Lightning-AI/lightning/pull/16681))
 
+- Removed the unused `lightning.pytorch.utilities.supporters.{SharedCycleIteratorState,CombinedLoaderIterator}` classes ([#16714](https://github.com/Lightning-AI/lightning/pull/16714))
```
[the remark this answers is not in the document]
Comment (carmocca) [reply src/lightning/pytorch/CHANGELOG.md:227]: I meant that we are removing it here because it would be unused after this refactor

**Pull request merged**